### ISSUE TITLE
[Feature request]: Android-like permissions

### ISSUE THREAD

**Issue by ghost** [flatpak/flatpak#5304]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a feature request that matches the one I want to file, without success.

### Suggestion

Most flatpaks grant home directory access by default, making it simple to escalate privileges (for example, by adding a command to '.bashrc').
Every app should, in my opinion, be devoid of permissions by default and prompt the user for them when it launches for the first time, like
```
This app wants to access
Microphone
Camera
User files

Allow/don't allow
```
There may also be integrations with Gnome and KDE with a menu that defines which permissions are enabled by default, which are blocked by default, and for which ask. This would greatly increase security and let users know what privileges the app has.

**Comment by imhemish** [comment]
Flatpak already has permissions and informs users of permissions in command line form, this is something that needs to be worked in GNOME Software and KDE Discover, and maybe xdg-desktop-portal too. So, I think this issue should be closed.

**Comment by Ultrabenosaurus** [comment]
> Flatpak already has permissions and informs users of permissions in command line form, this is something that needs to be worked in GNOME Software and KDE Discover, and maybe xdg-desktop-portal too. So, I think this issue should be closed.

If Flatpak already has the permissions system and the functionality to notify users via command line, this needs expanding to notify visually for GUI software as well. If Flatpak is to support GUI programs on desktop installations, as it does, then I don't feel its right to sweep this under the rug with "_well it works if you use the command line instead_". 

Relying on each individual OS to monitor Flatpak output and present it to the user, correctly and fully, is a sure-fire way to ensure it doesn't happen at best or gets implemented poorly and hurts Flatpak's reputation at worst, and I don't know see why Flatpak developers would want to completely give up control like that.

Flatpak should present Flatpak output properly to its users.

**Comment by imhemish** [comment]
> > Flatpak already has permissions and informs users of permissions in command line form, this is something that needs to be worked in GNOME Software and KDE Discover, and maybe xdg-desktop-portal too. So, I think this issue should be closed.
> 
> If Flatpak already has the permissions system and the functionality to notify users via command line, this needs expanding to notify visually for GUI software as well. If Flatpak is to support GUI programs on desktop installations, as it does, then I don't feel its right to sweep this under the rug with "_well it works if you use the command line instead_".
> 
> Relying on each individual OS to monitor Flatpak output and present it to the user, correctly and fully, is a sure-fire way to ensure it doesn't happen at best or gets implemented poorly and hurts Flatpak's reputation at worst, and I don't know see why Flatpak developers would want to completely give up control like that.
> 
> Flatpak should present Flatpak output properly to its users.

I mean this is out of scope of flatpak itself, flatpak itself is already providing the implementation, you just need something to wrap it up in GUI, and that is what portals do. Like file picking in flatpak, how does it work, flatpak contacts the portals to ask the desktop to do it. and desktops ship with their portal implementations, how they want to show it, i.e. how they want in their GUI. you just said 'relying on each individual os to monitor flatpak output and present it to user', noone monitors the output, they contact the portals, or the flatpak service for it directly, they dont 'watch' output, there are better ways to do without 'watching' the output. Flatpak itself should not implement the prompts shown to user, suppose if they do implement, they have to make some decisions, what toolkit to use for the prompts, so they rely on portals and desktop implementations to do it instead.

> Flatpak should present Flatpak output property to its users.

It already displays it 'properly', and flatpak CLI is not meant to be used by normal people who want to use GUI (sure they can use). It is the duty of software apps like GNOME Software and KDE Discover, to use this information provided by flatpak (not through seeing the output of commands, but rather using the flatpak service API). So, what you expect should be filed into issues of portals and software centers and not flatpak itself. Flatpak has done its work and already provides the needed information.

For example, GNOME Software recently got a feature to display if the app being installed is proprietary or not. Where is this information coming from? from a thing called tags, which are being specified in flatpak package of the app. flatpak service 'provides' this information, which is used by gnome software to show that this app is proprietary.

**Comment by Ultrabenosaurus** [comment]
> you just said 'relying on each individual os to monitor flatpak output and present it to user', noone monitors the output, they contact the portals, or the flatpak service for it directly, they dont 'watch' output, there are better ways to do without 'watching' the output.

I didn't use exactly the right terminology, sorry. My point was still clear, though, I believe - whether directly monitoring some CLI output in the background or using the service APIs or whatever, you're saying it should be down to each OS to specifically account for, integrate with, and provide functionality specifically for Flatpak's permissions system and then keep up-to-date with any changes Flatpak make. I am disagreeing with that take, and posit that it's reasonable for Flatpak to handle its own al;ready-existing features in a way that users can see and interact with via GUI if it's something important like app permissions.

> I mean this is out of scope of flatpak itself, flatpak itself is already providing the implementation, you just need something to wrap it up in GUI, and that is what portals do. Like file picking in flatpak, how does it work, flatpak contacts the portals to ask the desktop to do it. and desktops ship with their portal implementations, how they want to show it, i.e. how they want in their GUI.

> Flatpak itself should not implement the prompts shown to user, suppose if they do implement, they have to make some decisions, what toolkit to use for the prompts, so they rely on portals and desktop implementations to do it instead.

File picking and app permissions are not comparable, in my mind. File picking, or actually accessing a camera, is a function of the OS while notifying of app permissions and / or requesting approval from the user is directly related to running the program \*before\* it calls the OS to interact with those things, which is entirely within the scope of Flatpak. If it wasn't in scope, Flatpak wouldn't have a permissions system at all, CLI or GUI or hidden or in any other possible format.

Flatpak packages the app and provides what it needs to run and interface with the OS, including a permissions system for the packaged software - so I don't see how you can claim it is "out of scope" for Flatpak to properly present and handle those permissions with the user.

> It already displays it 'properly', and flatpak CLI is not meant to be used by normal people who want to use GUI (sure they can use).

If a notable portion of users use Flatpak for GUI software (I don't know the stats but I'd imagine "notable portion" is a fair assumption) and Flatpak doesn't display \*any\* permissions-related stuff outside of CLI then I'd say - especially given the subject of this feature request specifically asking for an Android-like user-facing permissions system - that Flatpak does not display that information properly.

This is also rather contradictory. In your first comment you say "_Flatpak already has permissions and informs users of permissions in command line form_" but now you are saying that the "_Flatpak CLI is not meant to be used by normal people_". If the CLI which shows permissions is not meant to be used by normal people, then Flatpak \*does not \* inform users about permissions when running the app.

**Comment by orowith2os** [comment]
>  If a notable portion of users use Flatpak for GUI software (I don't know the stats but I'd imagine "notable portion" is a fair assumption) and Flatpak doesn't display *any* permissions-related stuff outside of CLI then I'd say - especially given the subject of this feature request specifically asking for an Android-like user-facing permissions system - that Flatpak does not display that information properly.

out of scope, it's not flatpak's job to make frontends show related information. The frontends should handle that, all the information needed is there.

**Comment by kekkoudesu** [comment]
> Flatpak already has permissions and informs users of permissions in command line form, this is something that needs to be worked in GNOME Software and KDE Discover, and maybe xdg-desktop-portal too. So, I think this issue should be closed.

The feature `ghost` was asking for is different to informing users of the permissions. Flatpak packages ship with permissions by default, without giving the user a chance to refuse a particular permission they don't want to surrender. Thus, it's very easy for a malicious package to ask for a lot of permissions to do malicious things and be immediately granted that access.

With Android, iOS, and macOS, the user is prompted when a program tries to do something that requires privileged access:

> **Calculator** wants to access your entire filesystem. Do you want to allow access?
> **Calculator** wants to access your network. Do you want to allow access?

At the moment, presumably because developers for Flatpak packages don't want to scare users, some Flatpak packages ship with limited permissions that make programs hard to use by default. I had this experience with Bottles. I didn't know that I had to move all my game files from `~/games` to `~/var/app/com.usebottles.bottles/data/bottles/name/drive_c/games` after creating the bottle for it to work. I just tried selecting executable files in Bottles and wondered why it didn't work. Eventually, I figured out I needed to grant access to the folder, which I managed with Flatseal (I just gave it full access to my home because that seemed most convenient).

If "Android-like permissions" were implemented and I tried to access an executable at `~/games/`, it would prompt me saying:

> **Bottles** wants to access your ~/games folder. Do you want to allow access?

This would be both more user-friendly (it took me half an hour and some coaching to figure out the folder issue) and more secure by default. Newbies to Flatpak packages like myself often struggle with setting the right permissions. [Lutris essentially bypasses the sandbox entirely by granting wide-ranging permissions](https://github.com/flathub/net.lutris.Lutris/issues/229#issuecomment-1093378129) because they don't want to shift the burden of managing permissions onto the user. Not every use case necessitates these permissions, but enough do that Lutris has made that decision for all users by default:

> Flatseal is not convenient. You have to 1) know it exists 2) know why you need to use it 3) know how to use it. Flatseal is a solution that is suitable for computer engineers, not the general public. We want to ship Lutris as a working product. Not something that is broken by default.

So I think it would help a lot usability-wise while greatly enhancing security if programs needed to ask for every permission. This model has worked very well for iOS, Android, and macOS.

**Comment by orowith2os** [comment]
> So I think it would help a lot usability-wise while greatly enhancing security if programs needed to ask for every permission. This model has worked very well for iOS, Android, and macOS.

Those are called Portals. We already have them. This issue is non-actionable.

**Comment by kekkoudesu** [comment]
> Those are called Portals. We already have them. This issue is non-actionable.

I'm not very familiar with all this, sorry. If the plan is for each desktop's portal to implement the filesystem/network access prompts eventually, then that's great. It doesn't make much sense to keep this issue open if there's nothing to do on the Flatpak side.

**Comment by orowith2os** [comment]
Support for USB device access, safely, from a sandbox, is being worked on in xdg-desktop-portal right now: https://github.com/flatpak/xdg-desktop-portal/pull/559

And that includes hotplugging support. It requires kernel patches to work fully, but it's usable without them otherwise.

**Comment by GeorgesStavracas** [comment]
@orowith2os is correct in their assessment that this issue is essentially what XDG Desktop Portal does. There are various missing APIs that would benefit the desktop if we had them, but the foundations are there, and the appropriate repository discuss new portals is [flatpak/xdg](https://github.com/flatpak/xdg-desktop-portal. Technically it's fixed then? :slightly_smiling_face:

**Comment by colleirose** [comment]
> Flatpak already has permissions and informs users of permissions in command line form

And if a user isn't tech literate enough to understand everything in the command line? If the user knows what they're being informed of but wants to deny some of the permissions anyway?

> this is something that needs to be worked in GNOME Software and KDE Discover, and maybe xdg-desktop-portal too. So, I think this issue should be closed.

I agree that some apps need it. That's why it should be granted individually. Take VS Code as an example. It needs full file system access because it's for editing all your files, but it would still work just fine if the user could choose to set up its sandbox so that it thinks certain files or folders don't exist (say, for example, .bashrc). Hence why this should be up to the user and not require another application.

> If Flatpak already has the permissions system and the functionality to notify users via command line, this needs expanding to notify visually for GUI software as well. If Flatpak is to support GUI programs on desktop installations, as it does, then I don't feel its right to sweep this under the rug with "well it works if you use the command line instead".
> 
> Relying on each individual OS to monitor Flatpak output and present it to the user, correctly and fully, is a sure-fire way to ensure it doesn't happen at best or gets implemented poorly and hurts Flatpak's reputation at worst, and I don't know see why Flatpak developers would want to completely give up control like that.
> 
> Flatpak should present Flatpak output properly to its users.

This is correct, but people use GitHub reactions like the Reddit upvote-downvote system lol

> I mean this is out of scope of flatpak itself, flatpak itself is already providing the implementation, you just need something to wrap it up in GUI, and that is what portals do. Like file picking in flatpak, how does it work, flatpak contacts the portals to ask the desktop to do it. and desktops ship with their portal implementations, how they want to show it, i.e. how they want in their GUI. you just said 'relying on each individual os to monitor flatpak output and present it to user', noone monitors the output, they contact the portals, or the flatpak service for it directly, they dont 'watch' output, there are better ways to do without 'watching' the output. Flatpak itself should not implement the prompts shown to user, suppose if they do implement, they have to make some decisions, what toolkit to use for the prompts, so they rely on portals and desktop implementations to do it instead.

This is not how Flatpak is marketed.

Take a look at https://www.flatpak.org/ : "The future of apps on Linux", https://github.com/flatpak/flatpak "Linux application sandboxing", "Flatpak is a system for building, distributing, and running sandboxed desktop applications on Linux." It should, therefore, be able to provide proper sandboxing and not expect a hundred different implementations of it to all fix its sandboxing.

> It already displays it 'properly', and flatpak CLI is not meant to be used by normal people who want to use GUI (sure they can use). It is the duty of software apps like GNOME Software and KDE Discover, to use this information provided by flatpak (not through seeing the output of commands, but rather using the flatpak service API). So, what you expect should be filed into issues of portals and software centers and not flatpak itself. Flatpak has done its work and already provides the needed information.

This is not about telling the user what permissions an app has. It is about letting the user deny specific permissions to an app. These are very different things.

> out of scope, it's not flatpak's job to make frontends show related information. The frontends should handle that, all the information needed is there.

Yet again a failure to actually read the issue. The issue does not request Flatpak to "show related information". It requests "Android-like permissions" and for apps to "be devoid of permissions by default and prompt the user for them when it launches for the first time"

> Those are called Portals. We already have them. This issue is non-actionable.

You ignored almost all of the what you're replying to.

> That's the idea, and this part already works to some degree which covers a lot of use cases at least for programs which are willing to use the portals, but exceptional cases can be dealt with by giving access to specific directories, at least as far as filesystem access goes.

Then Flatpak should not advertise itself as providing sandboxing because it is expecting many different third-party developers to do the sandboxing.

> The tricky part is that it's not actually a bug that can be just fixed because there's just simply no compatibility guarantee by design. Most programs can't have hotplugging in containers because there's no working udev in the containerized environment and only a few are willing to implement a fallback approach, and Flatpak likely won't have just hotplugging on its own without notification because most programs would simply not detect the device appearing due to the udev problem, although the ones rescanning devices could benefit from such a change. Kind of a chicken and egg problem which is why "device=all" is still the only working way in many cases even despite the known security risks.

Indeed I am aware this is a complex issue, as is everyone else in the thread. However, I do not find it realistic to believe that it is physically impossible to implement a form of sandboxing that isn't all-or-nothing, either access to every file on the system or no files at all. Yes, I am not an experienced software engineer or a professional, I'm a hobbyist and most of my experience is in web development, but I feel that it doesn't make sense to claim that it is actually *not possible* to implement a similar form of sandboxing to what exists on Android, a Linux-based operating system, especially when apparently it's possible for a third-party application to implement the requested sandboxing.

> Support for USB device access, safely, from a sandbox, is being worked on in xdg-desktop-portal right now: https://github.com/flatpak/xdg-desktop-portal/pull/559

Great. "Support for USB device access, safely, from a sandbox" is not the main request.

> orowith2os is correct in their assessment that this issue is essentially what XDG Desktop Portal does. There are various missing APIs that would benefit the desktop if we had them, but the foundations are there, and the appropriate repository discuss new portals is [flatpak/xdg](https://github.com/flatpak/xdg-desktop-portal. Technically it's fixed then? 🙂

No they aren't correct. They misunderstood what is being requested.
